src/core/UniswapV2.sol

```diff
@@ -62,93 +114,20 @@ abstract contract UniswapV2 is SettlerAbstract {
-                // retrieve the sellToken balance of the pool
-                mstore(0x00, ERC20_BALANCEOF_SELECTOR)
-                mstore(0x20, pool)
-                // ||ERC20_BALANCEOF_SELECTOR|pool|
-                if iszero(staticcall(gas(), sellToken, 0x1c, 0x24, 0x00, 0x20)) { bubbleRevert(ptr) }
-                if lt(returndatasize(), 0x20) { revert(0, 0) }
-                let bal := mload(0x00)
-
-                // determine real sellAmount by comparing pool's sellToken balance to reserve amount
-                if lt(bal, sellReserve) {
-                    mstore(0x00, 0x4e487b71) // selector for `Panic(uint256)`
-                    mstore(0x20, 0x11) // panic code for arithmetic underflow
-                    revert(0x1c, 0x24)
-                }
-                sellAmount := sub(bal, sellReserve)
-            }
-
-            // compute buyAmount based on sellAmount and reserves
-            let sellAmountWithFee := mul(sellAmount, sub(10000, feeBps))
-            buyAmount := div(mul(sellAmountWithFee, buyReserve), add(sellAmountWithFee, mul(sellReserve, 10000)))
-            // set up swap call selector and empty callback data
-            mstore(ptr, UNI_PAIR_SWAP_SELECTOR)
-            mstore(add(ptr, 0x80), 0x80) // offset to length of data
-            mstore(add(ptr, 0xa0), 0) // length of data
-
-            // set amount0Out and amount1Out
-            let swapCalldata := add(ptr, 0x1c)
-            {
-                // If `zeroForOne`, offset is 0x24, else 0x04
-                let offset := add(0x04, shl(5, zeroForOne))
-                mstore(add(swapCalldata, offset), buyAmount)
-                mstore(add(swapCalldata, xor(0x20, offset)), 0)
-            }
-
-            mstore(add(swapCalldata, 0x44), and(0xffffffffffffffffffffffffffffffffffffffff, recipient))
-            // ...||UNI_PAIR_SWAP_SELECTOR|amount0Out|amount1Out|recipient|data|
-
-            // perform swap at the pool sending bought tokens to the recipient
-            if iszero(call(gas(), pool, 0, swapCalldata, 0xa4, 0, 0)) { bubbleRevert(ptr) }
-
-            // restore FMP
-            mstore(0x40, ptr)
-
-            // revert with the return data from the most recent call
-            function bubbleRevert(p) {
-                returndatacopy(p, 0, returndatasize())
-                revert(p, returndatasize())
-            }
+        (uint256 sellReserve, uint256 buyReserve) = fastUniswapV2Pool.fastGetReserves(pool, zeroForOne);
+        if (sellAmount == 0 || sellTokenHasFee) {
+            uint256 bal = IERC20(sellToken).fastBalanceOf(pool);
+            sellAmount = bal - sellReserve;
+        }
+        unchecked {
+            uint256 sellAmountWithFee = sellAmount * (10000 - feeBps);
+            buyAmount = (sellAmountWithFee * buyReserve) / (sellAmountWithFee + sellReserve * 10000);
         }
+        pool.fastSwap(zeroForOne, buyAmount, recipient);
         if (buyAmount < minBuyAmount) {
```

Comment:
the slippage check can be reordered above the call to `fastSwap` for marginal gas gains in the sad path